niriPipe/utils/finder.py

```diff
@@ -0,0 +1,396 @@
+#
+#  You should have received             Vous devriez avoir reçu une
+#  a copy of the GNU Affero             copie de la Licence Générale
+#  General Public License along         Publique GNU Affero avec
+#  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+#  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+#                                       <http://www.gnu.org/licenses/>.
+#
+#
+# ***********************************************************************
+import logging
+import astropy.table
+from astroquery.cadc import Cadc
+import re
+import io
+from cadcutils import net
+from cadcdata import CadcDataClient
+
+
+class Finder:
+    """
+    Finds all data for a given NIRI stack.
+    """
+
+    def __init__(self, state):
+        self.state = state
+        self.logger = logging.getLogger('niriPipe.utils.finder')
+        try:
+            self._log_basic_constraints()
+        except KeyError as e:
+            logging.critical(
+                "Insufficient constraints provided; " +
+                "is the config file complete?")
+            raise e
+
+    def run(self):
+        """
+        Runs the datafinder.
+
+        self._find_objects() gets object frames based on information in
+        self.state, and gets additional metadata from those object frames
+        about the current stack (things like integration time, filter, etc.)
+        This metadata (encoded in self.state) is then used to find appropriate
+        calibrations. NIRI needs flat field frames, long darks (darks with
+        the same integration time as science frames), and optional short darks
+        (1 second darks used to generate a bad pixel mask).
+        """
+        return astropy.table.vstack([
+            self._find_objects(),
+            self._segment(self._find_flats()),
+            self._segment(self._find_longdarks()),
+            self._segment(self._find_shortdarks())
+        ])
+
+    def _log_basic_constraints(self):
+        """
+        Log constraints that must be provided for the program to work.
+        """
+        self.logger.debug("Stack name: {}".format(
+            self.state['current_stack']['obs_name']))
+
+        self.logger.debug("Min required object frames: {}".format(
+            self.state['config']['DATAFINDER']['min_objects']))
+
+        self.logger.debug("Min required flat frames: {}".format(
+            self.state['config']['DATAFINDER']['min_flats']))
+
+        self.logger.debug("Min required longdark frames: {}".format(
+            self.state['config']['DATAFINDER']['min_longdarks']))
+
+        self.logger.debug("Min required shortdark frames: {}".format(
+            self.state['config']['DATAFINDER']['min_shortdarks']))
+
+    def _find_objects(self):
+        """
+        Find OBJECT frames and set metadata used to find calibrations.
+
+        Should be called before any of the find_cal_* methods.
+        """
+        object_query = \
```

Comment:
I would probably define a constant that is the beginning of each query, that looks something like this:

```
QUERY_BEGINNING = "SELECT publisherID, productID, energy_bandpassName, " + \
            "time_bounds_lower, time_exposure, type, intent " + \
            "FROM caom2.Plane AS Plane " + \
            "JOIN caom2.Observation AS Observation " + \
            "ON Plane.obsID = Observation.obsID " + \
            "WHERE Observation.collection = 'GEMINI' " + \
            "AND Observation.instrument_name = 'NIRI' "
```
and then just use that constant as the beginning of the bits that are query-specific, so something like:

```
query = "{} AND Plane.dataProductType = 'image' " + \
            "AND Observation.observationID LIKE '{}%' ".format(
                QUERY_BEGINNING, self.state['current_stack']['obs_name']
 ```
 
You may have heard this type of "don't repeat yourself" comment from me before, and if you have a good reason for specifying each query, that's ok too.

Reply:
Done! Ha, I'll learn if you tell me enough times!